frontend/src/pages/GraphicalSelector/GraphicalSelector.jsx

```diff
@@ -76,17 +76,14 @@ const GraphicalSelector = () => {
   };
 
   const setupGraph = async () => {
-    const { structure } = (
-      await axios.get(`/programs/getStructure/${programCode}/${specs.join("+")}`)
+    const { courses: courseList } = (
+      await axios.get(`/programs/getStructureCourseList/${programCode}/${specs.join("+")}`)
     ).data;
-    const courseList = (
-      Object.values(structure)
-        .flatMap((specialisation) => Object.values(specialisation)
-          .filter((spec) => typeof spec === "object" && spec.courses && !spec.type.includes("rule"))
-          .flatMap((spec) => Object.keys(spec.courses)))
-        .filter((v, i, a) => a.indexOf(v) === i) // TODO: hack to make courseList unique
-    );
+
+    // TODO: Move this to the backend too
+    // should be a universal /programs/getGraphEdges/{programCode}/{specs}
```

Comment:
I'm a bit confused with the approach here (not neccesarly disagreeing it tho). We're you planning on updating the `/programs/getStructureCourseList/` endpoint to become `/programs/getGraphEdges/`? So, `/programs/getGraphEdges/` would handle both the logic of:

```
const courseList = (
      Object.values(structure)
        .flatMap((specialisation) => Object.values(specialisation)
          .filter((spec) => typeof spec === "object" && spec.courses && !spec.type.includes("rule"))
          .flatMap((spec) => Object.keys(spec.courses)))
        .filter((v, i, a) => a.indexOf(v) === i) // TODO: hack to make courseList unique
    );
```
and
```
const res = await Promise.all(courseList.map((c) => axios.get(`/courses/getPathFrom/${c}`).catch((e) => e)));
```
?

I was initially thinking without this line that the approach you're approach taken would the BE return a course list from the endpoint. And have another endpoint just linking the edges, given a course list? But now I'm confuse haha.

Reply:
In my mind, the eventual goal was to have a route
```
/programs/getGraph/{programCode}/{spec}
```
This would
- construct the course structure
- flatten the course structure to get a list of courses
- find the edges corresponding to the course list
 
```
return {
  "courses": List[str],
  "edges": List[Tuple(str, str)]
}
```

My rationale was that the FE would need the `course` list anyways since otherwise courses with no edges get skipped but, I wanted to avoid the FE needing to send back a course list to get the edges.

The most optimal thing I could think of was to have everything sent back in a single call.